File: frontend/src/types/project.ts
export interface Project {
  id: string;
  name: string;
  description: string | null;
  owner_id: number;
}

export interface ProjectSummary extends Project {
  total_vulnerabilities: number;
  last_scan: string | null;
  scanners: string[] | null;
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
}

export interface SeveritySummary {
  critical: number;
  high: number;
  medium: number;
  low: number;
  info: number;
}

export interface ProjectDetails {
  project_id: string;
  project_name: string;
  total_scans: number;
  total_findings: number;
  severity: SeveritySummary;
  tools: Record<string, number>;
  last_scan: string | null;
}

export interface ProjectCreateRequest {
  name: string;
  description: string;
}

export interface ProjectUpdateRequest {
  name: string;
  description: string;
}

export interface ProjectAccessAssignment { user_id: number }
export interface ProjectAccessUser { id: number; email: string; first_name: string; last_name: string; effective_role_names: string[] }
export interface ProjectAccessGroup { id: number; name: string; description: string | null; member_count: number; effective_role_names: string[] }
export interface ProjectAccessResponse {
  users: ProjectAccessUser[];
  assignments: ProjectAccessAssignment[];
  groups: ProjectAccessGroup[];
  group_ids: number[];
}

File: backend/app/schemas/project.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, EmailStr, Field
from typing import Optional
from uuid import UUID


class ProjectCreate(BaseModel):
    name: str
    description: Optional[str] = None


class ProjectUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class ProjectResponse(BaseModel):
    id: UUID
    name: str
    description: Optional[str]
    owner_id: int

    model_config = ConfigDict(from_attributes=True)


class ProjectListResponse(ProjectResponse):
    total_vulnerabilities: int
    last_scan: datetime | None
    scanners: list[str] | None
    critical: int
    high: int
    medium: int
    low: int
    info: int


class ProjectAccessAssignment(BaseModel):
    user_id: int

    model_config = ConfigDict(from_attributes=True)


class ProjectAccessUpdate(BaseModel):
    assignments: list[ProjectAccessAssignment] = Field(default_factory=list)
    group_ids: list[int] = Field(default_factory=list)


class ProjectAccessUser(BaseModel):
    id: int
    email: EmailStr
    first_name: str
    last_name: str
    effective_role_names: list[str] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class ProjectAccessGroup(BaseModel):
    id: int
    name: str
    description: str | None
    member_count: int
    effective_role_names: list[str] = Field(default_factory=list)


class ProjectAccessResponse(BaseModel):
    users: list[ProjectAccessUser]
    assignments: list[ProjectAccessAssignment]
    groups: list[ProjectAccessGroup]
    group_ids: list[int]
